In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.grid_search import GridSearchCV
from sklearn.linear_model import LogisticRegression

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)


## Reading Avg w2v vectors(300-D)l

In [2]:
avg_w2v =pd.read_pickle("\\Users\\AKSHAY\\Desktop\\ML python course\\Amazon Project\\Logistic Regression\\W2V Vectors\\Avg W2V Vectors\\avg_w2v_vec_300")
print(avg_w2v.shape)
avg_w2v.head()

(122109, 301)


0         1         2         3         4         5         6    \
0  0.000259 -0.000116  0.000290 -0.000109  0.000132  0.000120  0.000027   
1  0.000189  0.000059  0.000166  0.000056 -0.000087  0.000035 -0.000047   
2 -0.000041 -0.000163  0.000031 -0.000424 -0.000413  0.000072  0.000474   
3  0.000260 -0.000109  0.000150 -0.000249 -0.000043 -0.000044  0.000222   
4  0.000038  0.000340 -0.000279  0.000020 -0.000101 -0.000203 -0.000285   

        7         8         9   ...        291       292       293       294  \
0  0.000366  0.000118  0.000003 ...   0.000197 -0.000364  0.000317  0.000149   
1  0.000082 -0.000092 -0.000212 ...   0.000141 -0.000036  0.000123  0.000098   
2 -0.000154 -0.000171  0.000377 ...  -0.000183  0.000082 -0.000136  0.000131   
3 -0.000008 -0.000014  0.000081 ...  -0.000005  0.000003  0.000053  0.000074   
4  0.000061  0.000015  0.000115 ...  -0.000407 -0.000198 -0.000115 -0.000093   

        295       296       297       298       299  300  
0  0.000193  0.000358  0.000094  0.000235 -0.000202    1  
1  0.000590 -0.000339  0.000291  0.000091 -0.000018    1  
2 -0.000009 -0.000278  0.000112 -0.000064  0.000302    0  
3 -0.000224  0.000161  0.000051 -0.000008  0.000007    1  
4 -0.000195 -0.000170 -0.000173 -0.000333 -0.000010    1  

[5 rows x 301 columns]

In [4]:
x =avg_w2v.iloc[:,:300]
y =avg_w2v.iloc[:,300]
print(x.shape)
y.shape

(122109, 300)


(122109,)

## Column Standardization

In [5]:
from sklearn.preprocessing import StandardScaler

In [6]:
a =StandardScaler()
x =a.fit_transform(x)

In [7]:
x_train,x_test,y_train,y_test =train_test_split(x,y,test_size =0.3)

In [8]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
y_test.shape

(85476, 300)
(36633, 300)
(85476,)


(36633,)

In [9]:
tuned_params = [{'C': [0.0001, 0.001, 0.01, 0.1, 1, 10, 100, 1000, 10000]}]

In [10]:
model = GridSearchCV(LogisticRegression(), tuned_params, scoring = 'accuracy')
model.fit(x_train, y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params={}, iid=True, n_jobs=1,
       param_grid=[{'C': [0.0001, 0.001, 0.01, 0.1, 1, 10, 100, 1000, 10000]}],
       pre_dispatch='2*n_jobs', refit=True, scoring='accuracy', verbose=0)

In [ ]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

In [12]:
print(model.best_estimator_)
print(model.score(x_test, y_test))

LogisticRegression(C=0.01, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)
0.777086233724


In [13]:
y_pred = model.best_estimator_.predict(x_test)

In [14]:
acc = accuracy_score(y_test, y_pred, normalize=True) * float(100)
acc

77.708623372369175

We can clearly see increasing dimenions the model performs better than 100-D,200-D as the chances of data becomes linearly separable in higher dimensions increases.So we get the best hyperplane as dimensions increases due to which model behaves better.

## Confusion Matrix

In [15]:
confusion_matrix(y_test,y_pred).T

array([[12951,  3967],
       [ 4199, 15516]], dtype=int64)

The confusion matrix of 300-D is better than other two

## L1 regularization and sparsity

In [16]:
clf = LogisticRegression(C=1, penalty = 'l1')
clf.fit(x_train, y_train)
w = clf.coef_    
print(np.count_nonzero(w))

300


This means 300 out of 300 are non zero.That means for regularization takes place.

In [17]:
clf = LogisticRegression(C=0.1, penalty = 'l1')
clf.fit(x_train, y_train)
w = clf.coef_
print(np.count_nonzero(w))

299


As we can see clearly as c decreases sparsity increases.This time 299 features out of 300  are non zero,means 1 features is removed.

In [18]:
clf = LogisticRegression(C=0.001, penalty = 'l1')
clf.fit(x_train, y_train)
w = clf.coef_
print(np.count_nonzero(w))

125


Now the 125 out of 300 are non zero.That means 175 are zero which are removed.

In [21]:
clf = LogisticRegression(C=0.00045, penalty = 'l1')
clf.fit(x_train, y_train)
w = clf.coef_
print(np.count_nonzero(w))

51


At c=0.00045 we are getting top 51 features and rest becomes zero.